scylla/src/transport/timestamp_generator.rs

```diff
@@ -0,0 +1,130 @@
+use std::{
+    sync::atomic::AtomicI64,
+    time::{SystemTime, UNIX_EPOCH},
+};
+
+use std::sync::atomic::Ordering;
+use std::sync::Mutex;
+use tokio::time::{Duration, Instant};
+use tracing::warn;
+
+/// Trait used to represent a timestamp generator
+pub trait TimestampGenerator: Send + Sync {
+    // This generates a new timestamp
+    fn next_timestamp(&self) -> i64;
+}
+
+/// Basic timestamp generator. Provides no guarantees, if system clock returns
+/// time before UNIX epoch it panics.
+#[derive(Default)]
+pub struct SimpleTimestampGenerator {}
+
+#[allow(dead_code)]
+impl SimpleTimestampGenerator {
+    pub fn new() -> Self {
+        SimpleTimestampGenerator {}
+    }
+}
+
+impl TimestampGenerator for SimpleTimestampGenerator {
+    fn next_timestamp(&self) -> i64 {
+        SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .unwrap()
+            .as_micros() as i64
+    }
+}
+
+/// Warning configuration for MonotonicTimestampGenerator
+pub struct MonotonicTimestampGeneratorCfg {
+    warning_threshold: Duration,
+    warning_interval: Duration,
+}
+
+/// Monotonic timestamp generator. Guarantees monotonicity of timestamps.
+/// If system clock will not provide an increased timestamp, then the timestamp will
+/// be artificially increased. If the config is provided and the clock skew is bigger than
+/// warning_threshold (by default 1 second), then the user will be warned about
+/// the skew repeatedly, with warning_interval provided in the settings (by default 1 second).
+/// Remember that this generator only guarantees monotonicity within one client!
+/// If you create multiple clients with MonotonicTimestampGenerators the monotonicity
+/// guarantee becomes void.
+pub struct MonotonicTimestampGenerator {
+    last: AtomicI64,
+    last_warning: Mutex<Instant>,
+    config: Option<MonotonicTimestampGeneratorCfg>,
+}
+
```

Comment:
I thought it would be more elegant, as IIRC timestamps are i64

Reply:
Yeah I forgot to delete this comment.
Neither of the types are perfect here. Protocol uses i64, but forbids negative values - another artifact of this protocol being designed with only Java in mind 🤦

Reply:
Actually  the AtomicU64 variant has the advantage that `Duration::from_micros` accepts u64, so I think it is the better choice in this case.
The trait should still use i64 - because that is what the protocol uses (even if it forbids negative values).